### DISCUSSION TITLE
WhenAnyPropertyChanged() did'nt work, because Transform()

### DISCUSSION THREAD

**Discussion by a44281071** [reactivemarbles/DynamicData#608] [Q&A] [answered]
I have:
    a source `SourceCache<UserInfo, string>`,
    a Items bind source `ObservableCollectionExtended<UserViewModel>`.

I connect() transform().

I listen Items collection, observe UserViewModel members change, use WhenAnyPropertyChanged(), and it didn't work.

test project : https://github.com/a44281071/TestDynamicData

**Comment by glennawatson** [comment]
Your `UserViewModel` doesn't implement `INotifyPropertyChanged` (and have the properties use it) so `WhenAnyPropertyChanged` does not know when properties have changed.

**Reply by a44281071** [reply]
nope，I implemented INotifyPropertyChanged by PropertyChanged.Fody.

**Reply by a44281071** [reply]
Same bug in `ObservableCacheEx.InnerJoin()` , resultSelector arg return a mapped Object same as Transform()

**Comment by RolandPheasant** [comment, marked as answer]
I see the issue, and there is a fundamental mistake in the sample code.   In reactive every subscription creates a new and independent stream .

As an example the code below

```cs
var chosen = itemsCache.Connect()
            .Transform(dd => new UserViewModel(dd))
            .AutoRefresh(p => p.IsCamOn)
            .Sort(SortExpressionComparer<UserViewModel>.Descending(e => e.IsCamOn))
            .Page(pager)
            .Do(change => PageParameters.Update(change.Response))
            .ObserveOnDispatcher()

chosen.WhenAnyPropertyChanged()
            .Subscribe(dd => Trace.TraceInformation("WhenAnyPropertyChanged() name = {0}", dd!.Info.Name));

 chosen.Bind(Items)
            .Subscribe();
```

will re-create 2 distinct RX streams, meaning that  each operator will be in invoked once per subscription. The implication is that chosen.WhenAnyPropertyChanged() and chosen.Bind(Items) will enforce all previous operators along the chain to run again. This means that the UserViewModel which is bound to is different to the UserViewModels where property changes are being listened to.


The answer is to share the connection using the rx Publish / Connect operator and pairs.

```cs
var chosen = itemsCache.Connect()
            .Transform(dd => new UserViewModel(dd))
            .AutoRefresh(p => p.IsCamOn)
            .Sort(SortExpressionComparer<UserViewModel>.Descending(e => e.IsCamOn))
            .Page(pager)
            .Do(change => PageParameters.Update(change.Response))
            .ObserveOnDispatcher()
            .Publish() //THE MAGIC IS HERE - means multiple subscribers share the output from this point onwards

chosen.WhenAnyPropertyChanged()
            .Subscribe(dd => Trace.TraceInformation("WhenAnyPropertyChanged() name = {0}", dd!.Info.Name));

 chosen.Bind(Items)
            .Subscribe();

// rx Connect() turns the sharing on.
// It is not dd Connect(). Ambiguous I know and it's a historic naming mistake.
var myConnectedDisposable =chosen.Connect() 
```

**Comment by a44281071** [comment]
Thanks!!! It's MAAAAAAAAAAAAAAAAAAAGIC!!!
